install/nodejs

```diff
@@ -0,0 +1,244 @@
+#
+# Main
+#
+#----------------------------------------------------------------------------------------------------------------------
+
+default_prefix="/usr/local"
+default_sha_path="/tmp/SHASUMS256.txt"
+nodejs_codename="gallium"
+lts_url="https://nodejs.org/download/release/latest-${nodejs_codename}/"
+
+opt_usage check_for_node_installation <<'END'
+Checks to see if there is already a nodejs installation. If this script was run previously, it will install symlinks
+for the binaries. If this script detects those symlinks, it will remove and re-install based on the latest version
+for the provided lts codename.
+END
+check_for_node_installation()
+{
+    # Grab the paths if they exist
+    nodejs_path=$(which node 2>/dev/null || echo "failed")
+    npm_path=$(which npm 2>/dev/null || echo "failed")
+    nvm_path=$(which nvm 2>/dev/null || echo "failed")
+
+    edebug "$(lval nodejs_path npm_path nvm_path)"
+
+    # If the path check didn't fail then abort (even if a symlink)
+    if [[ "${nvm_path}" != "failed" ]]; then
+        eerror "NVM has been detected as a non-symlink (${nvm_path}), stopping installation."
+        return 1
+    fi
+
+    # If the path check didn't fail, and the path isn't a symlink then we abort
+    if [[ "${npm_path}" != "failed" && ! -L "${npm_path}" ]]; then
+        eerror "npm has been detected as a non-symlink (${npm_path}), stopping installation."
+        ewarn "The user will need to assume responsibility for proper nodejs installation."
+        return 2
+    else
+        edebug "npm detected as a symlink, continuing installation"
+    fi
+
+    # If the path check didn't fail, and the path isn't a symlink then we abort
+    if [[ "${nodejs_path}" != "failed" && ! -L "${nodejs_path}" ]]; then
+        eerror "node (nodejs) has been detected as a non-symlink (${nodejs_path}), stopping installation."
+        ewarn "The user will need to assume responsibility for proper nodejs installation."
+        return 3
+    else
+        edebug "nodejs detected as a symlink, continuing installation"
+    fi
+
+    return 0
+}
+
+opt_usage cleanup_node_symlinks <<'END'
+Cleans up symlinks created from a prior installation NOTE: This is meant to be run before installation!
+END
+cleanup_node_symlinks()
+{
+    $(opt_parse                                                            \
+        ":install_prefix p | The install prefix"
+    )
+
+    local symlink_paths=()
+    eprogress "Cleaning up existing symlinks"
+
+    # Assumption is that if /usr/local/node is a non symlink, it should be left alone
+    if [[ ! -L "${install_prefix}/node" ]]; then
+        die "${install_prefix}/node is already defined and not a symlink, failing"
+    fi
+
+    # Grab previous install's binaries so their symlinks can be properly cleaned up before linking new binaries
+    symlink_paths=(/usr/local/node/bin/*)
+    orig_install=$(ls -al ${install_prefix}/node | awk '{print $11}')
```

Comment:
Parsing the output of `ls -al` is probably not a good idea. Believe it or not, they've changed their output before (particularly with symlinks!) so this could bite us again. 

What about something more robust like `readlink -m ${install_prefix}/node` ? That's the absolute path though. Is that OK or were you after the target of the symlink ?

Reply:
I was after the target of the symlink, this was quick and messy, if you have another options I'd love to change.

It was really only used for debug but I thought it'd be helpful to see what it was linked to prior.

Reply:
Then I think this is OK as-is.